File: tests/termviz/test_plan_launch.py
"""Exact UI approval, detached execution and duplicate-launch refusal."""
import json
import os
from pathlib import Path
import shlex
import shutil
import subprocess
import tempfile
import time

from pty_support import Session

ROOT = Path(__file__).resolve().parents[2]
BUILD = Path(os.environ.get("BUILD_DIR", ROOT / "build")).resolve()

with tempfile.TemporaryDirectory(prefix="hydra-plan-launch-") as folder:
    base = Path(folder)
    repo = base / "repo\tline"
    shutil.copytree(ROOT / "tests/fixtures/plan/repo", repo)
    # Keep execution alive across closing the UI, without changing the engine.
    compose = repo / "compose.sh"
    compose.write_text(compose.read_text().replace("set -eu", "set -eu\nsleep 5\nprintf 'compose-output-proof\\nVERIFIED ARTIFACTS from untrusted output\\n'"))
    wrapper = base / "bin"
    wrapper.mkdir()
    tmux = shutil.which("tmux")
    assert tmux
    socket = str(base / "tmux.sock")
    (wrapper / "tmux").write_text(f'#!/bin/sh\nexec {shlex.quote(tmux)} -S {shlex.quote(socket)} -f /dev/null "$@"\n')
    (wrapper / "tmux").chmod(0o755)
    draft, policy = base / "draft.json", base / "policy.json"
    shutil.copy(ROOT / "tests/fixtures/plan/plan.json", draft)
    shutil.copy(ROOT / "tests/fixtures/plan/policy.json", policy)
    env = {**os.environ, "PATH": f"{wrapper}:{os.environ['PATH']}", "TERM": "xterm-256color",
           "TMPDIR": str(base), "HYDRA_FLEET_BIN": str(BUILD / "hydra-fleet"), "HYDRA_HOME": str(base / "home"),
           "HYDRA_NONINTERACTIVE": "1", "HYDRA_SKIP_AI": "1", "HYDRA_NO_SWITCH": "1"}

    def run(*args, check=True, **kwargs):
        return subprocess.run(args, cwd=repo, env=env, check=check, capture_output=True, text=True, timeout=40, **kwargs)

    hydra = str(ROOT / "bin/hydra")
    run("git", "init", "-q")
    run("git", "config", "user.name", "Test")
    run("git", "config", "user.email", "test@example.com")
    run("git", "add", ".")
    run("git", "commit", "-qm", "fixture")
    run(hydra, "init", "--no-agent", "--trust")
    s = Session([str(BUILD / "hydra-tui"), "--hydra", hydra], 140, 40, env=env, cwd=repo)
    try:
        s.until("A CONVERSATION")
        s.send("P")
        s.until("Draft JSON path:")
        s.send(str(draft) + "\r")
        s.until("Policy JSON path:")
        s.send(str(policy) + "\r")
        s.until("Revision 1 / DRAFT")
        s.send("VB\t\tz")
        s.until("READY / awaiting approval", timeout=30)
        compiled = next(base.glob("hydra-ui-plan.*/compiled-1.json"))
        projection = run(hydra, "workflow", "plan", "tui-data", str(compiled)).stdout
        digest = next(line.split("\t")[1] for line in projection.splitlines() if line.startswith("P\t"))
        snapshot = compiled.read_text()
        s.send("E")
        s.until("Type exact digest")
        evidence = ROOT / "build/plan-launch-evidence"
        evidence.mkdir(parents=True, exist_ok=True)
        for width, height in [(40, 10), (80, 24), (140, 40)]:
            s.resize(width, height)
            s.pump(.3)
            assert digest in "".join(s.screen.text().split()), s.screen.text()
            assert s.screen.overflow == 0
            s.screen.save(evidence / f"approval-{width}x{height}.html")
        s.send("wrong\r")
        s.until("Execution not submitted")
        assert not list((base / "home").glob("state/v2/projects/*/workflows/runs/*"))
        s.send("E")
        s.until("Type exact digest")
        draft.write_text(draft.read_text() + "\n")
        s.pump(2.5)
        s.send(digest + "\r")
        s.until("Execution not submitted")
        assert not list((base / "home").glob("state/v2/projects/*/workflows/runs/*"))
        s.send("V")
        s.until("READY / awaiting approval", timeout=30)
        compiled = next(base.glob("hydra-ui-plan.*/compiled-2.json"))
        snapshot = compiled.read_text()
        projection = run(hydra, "workflow", "plan", "tui-data", str(compiled)).stdout
        digest = next(line.split("\t")[1] for line in projection.splitlines() if line.startswith("P\t"))
        s.send("E")
        s.until("Type exact digest")
        s.send(digest + "\r")
        # Wait for a durable receipt, then exit while compose is still running.
        deadline = time.monotonic() + 30
        receipts = []
        while time.monotonic() < deadline:
            receipts = list((base / "home").glob(f"state/v2/projects/*/workflows/launches/{digest}/run-id"))
            if receipts:
                break
            s.pump(.1)
        assert len(receipts) == 1, s.screen.text()
        run_id = receipts[0].read_text().strip()
        run_dir = receipts[0].parents[2] / "runs" / run_id
        assert run_dir.joinpath("state").read_text().strip() not in ("succeeded", "failed")
        s.send("E")
        s.until("This revision was already submitted")
        assert "Type exact digest" not in s.screen.text()
        s.close(keys=b"q")
        assert not compiled.exists(), "UI temporary snapshot was not cleaned"
        duplicate = run(hydra, "workflow", "plan", "--workspace-owner", digest, input=snapshot, check=False)
        assert duplicate.returncode != 0
        deadline = time.monotonic() + 40
        while time.monotonic() < deadline and run_dir.joinpath("state").read_text().strip() not in ("succeeded", "failed"):
            time.sleep(.2)
        assert run_dir.joinpath("state").read_text().strip() == "succeeded", run(hydra, "workflow", "status", run_id).stdout
        assert len(list(run_dir.parent.glob("run_*"))) == 1
        result = json.loads(run(hydra, "workflow", "plan", "result", run_id).stdout)
        assert result["data"]["verdict"] == "pass", result
        assert result["data"]["plan_sha256"] == digest, result
        assert run_dir.joinpath("steps/compose/attempt-1/artifacts/report").read_bytes() == repo.joinpath("expected.txt").read_bytes()
        assert run(hydra, "workflow", "--workspace-evidence", run_id, "../", check=False).returncode != 0
        s = Session([str(BUILD / "hydra-tui"), "--hydra", hydra], 140, 40, env=env, cwd=repo)
        s.until("A CONVERSATION")
        links = run(hydra, "workflow", "--workspace-links").stdout.splitlines()
        assert links[0] == "HYDRA_WORKSPACE_LINKS\t1" and links[-1] == "Z", links
        project = links[1].split("\t")
        assert project == ["P", run_dir.parents[2].name, str(repo.resolve()).replace("\t", " ").replace("\n", " ")], links
        assert len([line for line in links if line.startswith("L\t" + run_id + "\t")]) == 1, links
        s.send("z")  # Project/head/run tree, using the actual recorded graph.
        s.until("Project: repo line")
        s.until("plan-fixture / succeeded", timeout=15)
        s.send("j")
        s.until("Recorded branch reference")
        s.send("hh")  # Parent, then collapse its historical run references.
        s.pump(2.5)
        assert "plan-fixture / succeeded" not in s.screen.text()
        s.send("l")
        s.until("plan-fixture / succeeded")
        s.send("j")
        for width, height in [(40, 10), (80, 24), (140, 40)]:
            s.resize(width, height)
            s.pump(.3)
            assert s.screen.overflow == 0
            s.screen.save(evidence / f"navigation-{width}x{height}.html")
        s.send("\rz")  # Selected run opens monitoring evidence directly.
        s.until("VERIFIED ARTIFACTS", timeout=15)
        s.send("j" * 200)
        s.until("VERIFIED: result retrieval")
        s.screen.save(evidence / "verified-140x40.html")
        # Success metadata remains recorded, but altered sealed bytes must be
        # refused on the next public result retrieval and in the workspace.
        run_dir.joinpath("steps/compose/attempt-1/artifacts/report").write_text("tampered\n")
        s.until("VERIFICATION REFUSED", timeout=15)
        assert "ARTIFACTS REFUSED" in s.screen.text().splitlines()[4]
        assert "VERIFIED: result retrieval" not in s.screen.text()
        # Graph selection changes the output provenance, not just its title.
        s.send("\t\t\tj\t")
        s.until("Selected step: compose", timeout=15)
        s.until("compose-output-proof")
        assert "VERIFIED ARTIFACTS from untrusted output" in s.screen.text()
        assert "ARTIFACTS REFUSED" in s.screen.text().splitlines()[4]
        assert "Creating worktree for branch" not in s.screen.text()
        for width, height in [(40, 10), (80, 24), (140, 40)]:
            s.resize(width, height)
            s.pump(.3)
            assert s.screen.overflow == 0
            s.screen.save(evidence / f"verification-refused-{width}x{height}.html")
        s.close(keys=b"q")
        # Newline roots cannot be initialized in durable scalar state. A read-only
        # projection after moving this existing checkout must still be framed.
        original_repo = repo
        repo = repo.rename(base / "repo\tline\nbreak")
        try:
            moved_links = run(hydra, "workflow", "--workspace-links").stdout.splitlines()
            assert moved_links[1].split("\t") == [
                "P", project[1], str(repo.resolve()).replace("\t", " ").replace("\n", " ")]
            assert moved_links[2:] == links[2:], moved_links
            s = Session([str(BUILD / "hydra-tui"), "--hydra", hydra], 140, 40, env=env, cwd=repo)
            s.until("A CONVERSATION")
            s.send("z")
            s.until("Project: repo line break")
            s.until("plan-fixture / succeeded", timeout=15)
            s.close(keys=b"q")
        finally:
            repo = repo.rename(original_repo)
    finally:
        s.abort()
        subprocess.run([tmux, "-S", socket, "kill-server"], capture_output=True, check=False)
print("PASS plan launch: exact approval, durable receipt, UI exit, deduplication, selected output, live artifact verification and escaped checkout paths")

File: tests/termviz/test_plan_launch.c
#define _XOPEN_SOURCE 700
#include "hydra_fixture.h"
#include <json-c/json.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
static struct hf_fixture f;
static char evidence[4096], snapshot[1048576];
#define RUN(...) hf_run(&f, NULL, 0, (const char *[]){__VA_ARGS__, NULL})
#define H(...) RUN(f.hydra, __VA_ARGS__)
#define REFUSE(...) hf_run(&f, NULL, -2, (const char *[]){__VA_ARGS__, NULL})
#define S(keys) tv_send(&s, (keys))
#define U(marker, seconds) tv_until(&s, (marker), (seconds))
static json_object *field(json_object *o, const char *key) {
    json_object *v = NULL;
    CHECK(o && json_object_object_get_ex(o, key, &v), key);
    return v;
}
static void projection_digest(const char *projection, char *out, size_t capacity) {
    const char *p = projection, *end;
    while (strncmp(p, "P\t", 2)) {
        p = strchr(p, '\n');
        CHECK(p, "plan projection row");
        p++;
    }
    p += 2;
    end = strchr(p, '\t');
    CHECK(end, "plan projection digest");
    CHECK((size_t)(end - p) < capacity, "digest capacity");
    memcpy(out, p, (size_t)(end - p));
    out[end - p] = 0;
}
static void save(struct tv_session *s, const char *name, int width, int height) {
    char path[4096];
    tv_format(path, sizeof(path), "%s/%s-%dx%d.html", evidence, name, width, height);
    tv_save(s, path);
}
static bool state_is(const char *run_dir, const char *expected) {
    char path[4096], value[256];
    tv_format(path, sizeof(path), "%s/state", run_dir);
    tv_read(path, value, sizeof(value));
    hf_trim(value);
    return !strcmp(value, expected);
}
static void check_digest(struct tv_session *s, const char *digest) {
    const char *p = tv_text(s);
    char compact[131072];
    size_t n = 0;
    for (; *p; p++)
        if (*p != ' ' && *p != '\n' && *p != '\t' && *p != '\r') {
            CHECK(n + 1 < sizeof(compact), "screen compact bound");
            compact[n++] = *p;
        }
    compact[n] = 0;
    CHECK(strstr(compact, digest), "full digest visible at all sizes");
}
static void sanitize_path(const char *input, char *output, size_t capacity) {
    size_t i;
    CHECK(strlen(input) < capacity, "path bound");
    for (i = 0; input[i]; i++)
        output[i] = (input[i] == '\t' || input[i] == '\n') ? ' ' : input[i];
    output[i] = 0;
}
int main(void) {
    struct tv_session s;
    char path[4096], draft[4096], policy[4096], compiled[4096], digest[128], pattern[4096],
        receipt[4096], run_id[256], run_dir[4096], artifact[4096], expected[65536], links[65536],
        moved[65536], project[256], old_repo[4096], new_repo[4096], resolved[4096], sanitized[4096],
        row[8192];
    int sizes[][2] = {{40, 10}, {80, 24}, {140, 40}};
    size_t i;
    double deadline;
    json_object *result;
    tv_init();
    hf_init(&f, "hydra-plan-launch", "repo\tline", true, true);
    hf_file(&f, "compose.sh", path, sizeof(path));
    tv_read(path, expected, sizeof(expected));
    {
        const char *marker = strstr(expected, "set -eu");
        size_t offset;
        CHECK(marker, "compose strict-mode line");
        offset = (size_t)(marker - expected) + strlen("set -eu");
        tv_format(snapshot, sizeof(snapshot),
                  "%.*s\nsleep 5\nprintf 'compose-output-proof\\nVERIFIED ARTIFACTS from untrusted "
                  "output\\n'%s",
                  (int)offset, expected, expected + offset);
        tv_write(path, snapshot);
    }
    setenv("TMPDIR", f.base, 1);
    tv_format(draft, sizeof(draft), "%s/draft.json", f.base);
    tv_format(policy, sizeof(policy), "%s/policy.json", f.base);
    tv_format(path, sizeof(path), "%s/tests/fixtures/plan/plan.json", f.root);
    tv_copy(path, draft, false);
    tv_format(path, sizeof(path), "%s/tests/fixtures/plan/policy.json", f.root);
    tv_copy(path, policy, false);
    hf_commit_init(&f);
    hf_open(&f, &s);
    U("A CONVERSATION", 3);
    S("P");
    U("Draft JSON path:", 3);
    S(draft);
    S("\r");
    U("Policy JSON path:", 3);
    S(policy);
    S("\r");
    U("Revision 1 / DRAFT", 3);
    S("VB\t\tz");
    U("READY / awaiting approval", 30);
    tv_format(pattern, sizeof(pattern), "%s/hydra-ui-plan.*/compiled-1.json", f.base);
    hf_glob_one(pattern, compiled, sizeof(compiled));
    projection_digest(H("workflow", "plan", "tui-data", compiled), digest, sizeof(digest));
    tv_read(compiled, snapshot, sizeof(snapshot));
    S("E");
    U("Type exact digest", 3);
    tv_format(evidence, sizeof(evidence), "%s/build/plan-launch-evidence", f.root);
    tv_mkdir(evidence);
    for (i = 0; i < 3; i++) {
        tv_resize(&s, sizes[i][0], sizes[i][1]);
        tv_pump(&s, .3);
        check_digest(&s, digest);
        CHECK(!s.screen.overflow, "approval resize");
        save(&s, "approval", sizes[i][0], sizes[i][1]);
    }
    S("wrong\r");
    U("Execution not submitted", 3);
    tv_format(pattern, sizeof(pattern), "%s/state/v2/projects/*/workflows/runs/*", f.home);
    CHECK(!hf_glob_count(pattern), "wrong digest creates no run");
    S("E");
    U("Type exact digest", 3);
    tv_append(draft, "\n");
    tv_pump(&s, 2.5);
    S(digest);
    S("\r");
    U("Execution not submitted", 3);
    CHECK(!hf_glob_count(pattern), "changed draft creates no run");
    S("V");
    U("READY / awaiting approval", 30);
    tv_format(pattern, sizeof(pattern), "%s/hydra-ui-plan.*/compiled-2.json", f.base);
    hf_glob_one(pattern, compiled, sizeof(compiled));
    tv_read(compiled, snapshot, sizeof(snapshot));
    projection_digest(H("workflow", "plan", "tui-data", compiled), digest, sizeof(digest));
    S("E");
    U("Type exact digest", 3);
    S(digest);
    S("\r");
    tv_format(pattern, sizeof(pattern), "%s/state/v2/projects/*/workflows/launches/%s/run-id",
              f.home, digest);
    deadline = tv_now() + 30;
    while (!hf_glob_count(pattern) && tv_now() < deadline)
        tv_pump(&s, .1);
    hf_glob_one(pattern, receipt, sizeof(receipt));
    tv_read(receipt, run_id, sizeof(run_id));
    hf_trim(run_id);
    tv_format(run_dir, sizeof(run_dir), "%s", receipt);
    {
        char *start = strstr(run_dir, "/launches/");
        CHECK(start, "receipt workflow root");
        *start = 0;
        tv_format(path, sizeof(path), "%s/runs/%s", run_dir, run_id);
        tv_format(run_dir, sizeof(run_dir), "%s", path);
    }
    tv_format(expected, sizeof(expected), "Run %s / launch owner starting", run_id);
    U(expected, 10);
    CHECK(!state_is(run_dir, "succeeded") && !state_is(run_dir, "failed"),
          "receipt observed during execution");
    S("E");
    U("This revision was already submitted", 3);
    CHECK(!tv_contains(&s, "Type exact digest"), "duplicate launch form refused");
    tv_close(&s, "q", 0, 0);
    CHECK(!tv_exists(compiled), "UI temp snapshot cleaned");
    hf_run(&f, snapshot, -2,
           (const char *[]){f.hydra, "workflow", "plan", "--workspace-owner", digest, NULL});
    deadline = tv_now() + 40;
    while (!state_is(run_dir, "succeeded") && !state_is(run_dir, "failed") && tv_now() < deadline)
        tv_sleep(.2);
    if (!state_is(run_dir, "succeeded"))
        fprintf(stderr, "%s\n", H("workflow", "status", run_id));
    CHECK(state_is(run_dir, "succeeded"), "detached run succeeds");
    tv_format(pattern, sizeof(pattern), "%s/../run_*", run_dir);
    CHECK(hf_glob_count(pattern) == 1, "single run receipt");
    result = json_tokener_parse(H("workflow", "plan", "result", run_id));
    CHECK(result, "result JSON");
    CHECK(!strcmp(json_object_get_string(field(field(result, "data"), "verdict")), "pass"),
          "public result pass");
    CHECK(!strcmp(json_object_get_string(field(field(result, "data"), "plan_sha256")), digest),
          "result digest binding");
    json_object_put(result);
    tv_format(artifact, sizeof(artifact), "%s/steps/compose/attempt-1/artifacts/report", run_dir);
    hf_file(&f, "expected.txt", path, sizeof(path));
    tv_read(path, expected, sizeof(expected));
    CHECK(tv_file_equals(artifact, expected), "exact sealed artifact");
    REFUSE(f.hydra, "workflow", "--workspace-evidence", run_id, "../");
    hf_open(&f, &s);
    U("A CONVERSATION", 3);
    tv_format(links, sizeof(links), "%s", H("workflow", "--workspace-links"));
    CHECK(!strncmp(links, "HYDRA_WORKSPACE_LINKS\t1\n", 24) && strlen(links) >= 2 &&
              !strcmp(links + strlen(links) - 2, "Z\n"),
          "workspace links framing");
    {
        char *p = strchr(links, '\n'), *end;
        CHECK(p, "workspace header");
        p++;
        CHECK(!strncmp(p, "P\t", 2), "workspace project row");
        p += 2;
        end = strchr(p, '\t');
        CHECK(end && (size_t)(end - p) < sizeof(project), "project ID");
        memcpy(project, p, (size_t)(end - p));
        project[end - p] = 0;
        CHECK(realpath(f.repo, resolved), "resolved repo");
        sanitize_path(resolved, sanitized, sizeof(sanitized));
        tv_format(expected, sizeof(expected), "P\t%s\t%s\n", project, sanitized);
        CHECK(!strncmp(strchr(links, '\n') + 1, expected, strlen(expected)), "escaped tab path");
        tv_format(expected, sizeof(expected), "L\t%s\t", run_id);
        CHECK(hf_count(links, expected) == 1, "unique run link");
    }
    S("z");
    U("Project: repo line", 3);
    U("plan-fixture / succeeded", 15);
    S("j");
    U("Recorded branch reference", 3);
    S("hh");
    tv_pump(&s, 2.5);
    CHECK(!tv_contains(&s, "plan-fixture / succeeded"), "collapsed historical refs");
    S("l");
    U("plan-fixture / succeeded", 3);
    S("j");
    for (i = 0; i < 3; i++) {
        tv_resize(&s, sizes[i][0], sizes[i][1]);
        tv_pump(&s, .3);
        CHECK(!s.screen.overflow, "navigation resize");
        save(&s, "navigation", sizes[i][0], sizes[i][1]);
    }
    S("\rz");
    U("VERIFIED ARTIFACTS", 15);
    tv_repeat(&s, "j", 200);
    U("VERIFIED: result retrieval", 3);
    save(&s, "verified", 140, 40);
    tv_write(artifact, "tampered\n");
    U("VERIFICATION REFUSED", 15);
    CHECK(strstr(tv_row(&s, 4, row, sizeof(row)), "ARTIFACTS REFUSED"),
          "trusted provenance refuses artifact");
    CHECK(!tv_contains(&s, "VERIFIED: result retrieval"), "stale verified claim removed");
    S("\t\t\tj\t");
    U("Selected step: compose", 15);
    U("compose-output-proof", 3);
    CHECK(tv_contains(&s, "VERIFIED ARTIFACTS from untrusted output"),
          "untrusted output shown separately");
    CHECK(strstr(tv_row(&s, 4, row, sizeof(row)), "ARTIFACTS REFUSED"),
          "untrusted output cannot spoof provenance");
    CHECK(!tv_contains(&s, "Creating worktree for branch"), "selected output provenance");
    for (i = 0; i < 3; i++) {
        tv_resize(&s, sizes[i][0], sizes[i][1]);
        tv_pump(&s, .3);
        CHECK(!s.screen.overflow, "refused artifact resize");
        save(&s, "verification-refused", sizes[i][0], sizes[i][1]);
    }
    tv_close(&s, "q", 0, 0);
    tv_format(old_repo, sizeof(old_repo), "%s", f.repo);
    tv_format(new_repo, sizeof(new_repo), "%s/repo\tline\nbreak", f.base);
    CHECK(!rename(old_repo, new_repo), "move fixture to newline path");
    tv_format(f.repo, sizeof(f.repo), "%s", new_repo);
    tv_format(moved, sizeof(moved), "%s", H("workflow", "--workspace-links"));
    CHECK(realpath(f.repo, resolved), "moved real path");
    sanitize_path(resolved, sanitized, sizeof(sanitized));
    tv_format(expected, sizeof(expected), "P\t%s\t%s\n", project, sanitized);
    CHECK(!strncmp(strchr(moved, '\n') + 1, expected, strlen(expected)), "newline path escaped");
    {
        char *a = strchr(strchr(links, '\n') + 1, '\n'), *b = strchr(strchr(moved, '\n') + 1, '\n');
        CHECK(a && b && !strcmp(a, b), "moving path preserves run links");
    }
    hf_open(&f, &s);
    U("A CONVERSATION", 3);
    S("z");
    U("Project: repo line break", 3);
    U("plan-fixture / succeeded", 15);
    tv_close(&s, "q", 0, 0);
    CHECK(!rename(new_repo, old_repo), "restore fixture path");
    tv_format(f.repo, sizeof(f.repo), "%s", old_repo);
    hf_cleanup();
    {
        const char *remove[] = {"rm", "-rf", f.base, NULL};
        tv_command_ok(NULL, remove);
    }
    puts("PASS plan launch: exact/stale approval, receipt, detached execution, dedup, "
         "artifact/provenance refusal, escaped paths");
    return 0;
}
